bazel: take into account system wide bazelrc configuration if it exists

guibou opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

closes #70911

This was extracted from https://github.com/NixOS/nixpkgs/pull/73309 to respect separation of concern.

Edit: after a few discussion with @flokli, this PR also refactors a bit of the bazelrc configuration handling:

- the new bazelrc file is generated in its own nix store path. That's cleaner.
- the replacement in the bazel source code is done using a patch, this way a regression will be catched earlier on.
- the wrap of bazel was replaced by entries in the bazelrc file

cc @flokli 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @flokli

flokli commented:
We currently abuse bazels `/etc/bazel/bazelrc` feature to "configure" nixpkgs bazel.

We patch bazel to read from `$out/etc/bazelrc` (causing #70911), and then append some config statements in there while building bazel.

We also use `wrapProgram` to append some options to the bazel binary directly: `--server_javabase="${runJdk}"`.

I think we use `$out/etc/bazelrc` simply because some of the arguments would need to be passed to the binary depending on `$argv[1]`, which requires a custom wrapper.

I think the cleanest way forward would probably be to simply patch the default locations hardcoded into the source code. Maybe there's even a way to configure those during runtime, I haven't really checked.

cc @aherrmann

flokli commented:
Can this PR replace the `wrapProgram "$out/bin/bazel" --add-flags --server_javabase="${runJdk}"` with adding that flag to the bazelrc as well, so we at least only use one way to configure the nixpkgs-style bazel?

guibou committed:
bazel: move configuration file to a proper nix store entry

Easier to read and maintain

guibou committed:
bazel: remove wrapProgram and move it to system configuration

This way, custom configurations are all at the same place. Also, user
may override these settings using the command line.

guibou committed:
bazel: close #70911 by conditionally loading system wide configuration

Some system may setup global informations (such as bazel cache location)
in /etc/bazel.bazelrc

guibou committed:
bazel: bazelRC path set by a patch

In bazel source code change in the future, patch will fail and we'll
detect the regression sooner.

flokli commented:
@GrahamcOfBorg build bazel bazel.tests
